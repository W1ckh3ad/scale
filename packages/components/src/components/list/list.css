.list {
  padding-left: 0;
}

.list--nested {
  margin-top: 0;
  margin-bottom: 0;
  padding-left: var(--scl-spacing-8);
}
