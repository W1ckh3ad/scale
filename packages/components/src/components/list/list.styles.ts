import { JssStyle } from 'jss';

export const styles: JssStyle = {
  list: {
    paddingLeft: ({ spacing }) => spacing.flow,
  },
  // WIP This only works *without* `scale-list-item`
  'list--variant-ordered': {
    // listStyle: 'none',
    // paddingLeft: 0,
    // counterReset: 'scale-counter',
    '& ::slotted(li)': {
      counterIncrement: 'scale-counter',
    },
    '& ::slotted(li)::before': {
      content: 'counter(scale-counter) "."',
      fontWeight: ({ font }) => font.weight_bold,
      paddingRight: ({ spacing }) => spacing.inline,
    },
  },
};
