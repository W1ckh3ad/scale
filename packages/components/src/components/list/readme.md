# scale-list

<!-- Auto Generated Below -->


## Properties

| Property  | Attribute | Description                           | Type      | Default     |
| --------- | --------- | ------------------------------------- | --------- | ----------- |
| `ordered` | `ordered` | (optional) Make the list ordered (ol) | `boolean` | `false`     |
| `styles`  | `styles`  | (optional) Injected jss styles        | `any`     | `undefined` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
