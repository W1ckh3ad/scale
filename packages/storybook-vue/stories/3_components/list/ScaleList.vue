<script>
export default {
  props: {
    ordered: { type: Boolean },
    styles: { type: Object }
  },
  render() {
    return this.$slots.default
  }
}
</script>