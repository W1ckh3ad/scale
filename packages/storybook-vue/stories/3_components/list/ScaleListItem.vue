<template>
  <scale-list-item
    :icon="icon"
    :icon-size="iconSize"
    :type="type"
    :styles="styles"
  >
    <slot></slot>
  </scale-list-item>
</template>

<script>
export default {
  props: {
    icon: { type: String },
    iconSize: { type: Number, default: 16 },
    type: { type: String },
    styles: { type: Object }
  }
}
</script>