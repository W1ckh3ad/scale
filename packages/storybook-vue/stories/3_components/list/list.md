# Text List

List items help users determine hierarchies in related content.

## General

### When to Use

Use lists only for displaying information.

## Basic Types

### Ordered Lists

Use ordered lists to display data that has a serial relationship.

![Image Name](/assets/3_components/text-list/image-20200810132442835.png)

### Unordered Lists

Use unordered lists to display data with no specifc serial succession.

![Image Name](/assets/3_components/text-list/image-20200810132054388.png)

### Definition Lists

Use definition lists to display terms along with their definition.

![Image Name](/assets/3_components/text-list/image-20200807134929248.png)

## Code



