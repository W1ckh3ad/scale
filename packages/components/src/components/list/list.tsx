import { Component, h, Prop, Host, Element, Watch } from '@stencil/core';
import { CssClassMap } from '../../utils/utils';
import classNames from 'classnames';
import { styles } from './list.styles';
import { CssInJs } from '../../utils/css-in-js';
import { StyleSheet } from 'jss';
import Base from '../../utils/base-interface';

/**
 * @see https://github.com/carbon-design-system/carbon-web-components/tree/master/src/components/list
 */
@Component({
  tag: 'scale-list',
  shadow: true,
})
export class List implements Base {
  isNested: boolean = false;

  @Element() el: HTMLElement;

  /** (optional) Make the list ordered (ol) */
  @Prop() ordered?: boolean = false;
  /** (optional) Injected jss styles */
  @Prop() styles?: any;
  /** decorator Jss stylesheet */
  @CssInJs('List', styles) stylesheet: StyleSheet;

  @Watch('ordered')
  orderedChanged(newValue) {
    this.propagatePropsToChildren(newValue);
  }

  componentWillUpdate() {}
  componentDidUnload() {}

  componentDidLoad() {
    this.propagatePropsToChildren(this.ordered);
  }

  connectedCallback() {
    this.isNested = this.el.closest('scale-list-item') != null;

    if (this.isNested) {
      this.el.setAttribute('slot', 'nested');
    } else {
      this.el.removeAttribute('slot');
    }
  }

  propagatePropsToChildren(ordered: boolean) {
    const items = Array.from(this.el.children).filter(child =>
      child.matches('scale-list-item')
    );
    (items as HTMLScaleListItemElement[]).forEach((item, index) => {
      item.ordered = ordered;
      item.index = index + 1;
    });
  }

  render() {
    const Tag = this.ordered ? 'ol' : 'ul';

    return (
      <Host>
        <style>{this.stylesheet.toString()}</style>
        <Tag class={this.getCssClassMap()}>
          <slot />
        </Tag>
      </Host>
    );
  }

  getCssClassMap(): CssClassMap {
    const { classes } = this.stylesheet;
    return classNames(
      classes.list,
      this.ordered && classes['list--type-ordered'],
      this.isNested && classes['list--nested']
    );
  }
}
