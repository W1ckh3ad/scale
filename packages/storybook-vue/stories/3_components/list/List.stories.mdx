import {
  Meta,
  Story,
  Props,
  Preview,
  Description,
} from "@storybook/addon-docs/blocks";
import ScaleList from "./ScaleList.vue";
import ScaleListItem from "./ScaleListItem.vue";
import description from "./list.md";
const ICON =
  "M6.79280818,20.9472559 C6.4228823,20.577241 6.40348086,19.9894966 6.73455424,19.5966427 L6.79297827,19.5330423 L14.206547,12.1215935 L6.79289324,4.7071068 C6.42292284,4.33713642 6.40345071,3.74939435 6.73447683,3.35650062 L6.7928932,3.29289324 C7.16286358,2.92292284 7.75060565,2.90345071 8.14349938,3.23447683 L8.20710676,3.2928932 L17.0353925,12.1211785 L8.20702173,20.9474259 C7.81645047,21.3379033 7.1832855,21.3378271 6.79280818,20.9472559 Z";

<Meta
  title="Components/List"
  component={ScaleList}
  subcomponents={{ ScaleListItem }}
  argTypes={{
    variant: {
      control: { type: "radio", options: ["unordered", "ordered"] },
    },
    styles: {
      control: { disable: true },
    },
  }}
/>

export const Template = (args, { argTypes }) => ({
  components: { ScaleList },
  props: {
    label: String,
    icon: String,
    iconSize: Number,
    type: String,
    ...ScaleList.props,
  },
  template: `
    <scale-list :variant="variant">
      <scale-list-item :type="type" :icon="icon" :icon-size="iconSize">Eins</scale-list-item>
      <scale-list-item :type="type" :icon="icon" :icon-size="iconSize">Zwei</scale-list-item>
      <scale-list-item :type="type" :icon="icon" :icon-size="iconSize">Drei</scale-list-item>
    </scale-list>
  `,
});

<Description>{description}</Description>

<Preview withSource="none">
  <Story name="Standard" args={{ variant: "unordered" }}>
    {Template.bind({})}
  </Story>
</Preview>

<Props story="Standard" />

```html
<scale-list variant="unordered">
  <scale-list-item>Eins</scale-list-item>
  <scale-list-item>Zwei</scale-list-item>
  <scale-list-item>Drei</scale-list-item>
</scale-list>
```

## With Custom Type

<Preview withSource="none">
  <Story
    name="With Custom Type"
    args={{
      type: "lower-roman",
    }}
  >
    {Template.bind({})}
  </Story>
</Preview>

```html
<scale-list>
  <scale-list-item type="lower-roman">Eins</scale-list-item>
  <scale-list-item type="lower-roman">Zwei</scale-list-item>
  <scale-list-item type="lower-roman">Drei</scale-list-item>
</scale-list>
```

## With Custom Icon

<Preview withSource="none">
  <Story
    name="With Custom Icon"
    args={{
      icon: ICON,
      iconSize: 16,
    }}
  >
    {Template.bind({})}
  </Story>
</Preview>

```html
<scale-list>
  <scale-list-item icon="SVG-PATH" icon-size="16">Eins</scale-list-item>
  <scale-list-item icon="SVG-PATH" icon-size="16">Zwei</scale-list-item>
  <scale-list-item icon="SVG-PATH" icon-size="16">Drei</scale-list-item>
</scale-list>
```
