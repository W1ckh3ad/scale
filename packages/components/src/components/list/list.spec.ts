import { newSpecPage } from '@stencil/core/testing';
import { List } from './list';

describe('List', () => {
  let element;
  beforeEach(async () => {
    element = new List();
  });

  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [List],
      html: `<scale-list>default</scale-list>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should handle css classes', () => {
    element.ordered = true;
    expect(element.getCssClassMap()).toContain('list--type-ordered');
  });
});
