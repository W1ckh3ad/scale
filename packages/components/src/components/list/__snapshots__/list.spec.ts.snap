// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should match snapshot 1`] = `
<scale-list>
  <mock:shadow-root>
    <style>
      .list-0-0-4 {
  padding-left: 0;
}
.list--nested-0-0-5 {
  margin-top: 0;
  margin-bottom: 0;
  padding-left: 0.5rem;
}
.list--type-ordered-0-0-6 {}
    </style>
    <ul class="list-0-0-4">
      <slot></slot>
    </ul>
  </mock:shadow-root>
  default
</scale-list>
`;
